File: 인터넷 강의/python_ruby/Class_member/4.py
class Cal(object):
    
    _history = []

    def __init__(self, v1, v2):
        self.v1 = v1
        self.v2 = v2

    def add(self):
        result = self.v1 + self.v2
        Cal._history.append("add : %d + %d = %d" % (self.v1, self.v2, result))
        return result

    def subtract(self):
        result = self.v1 - self.v2
        Cal._history.append("subtract : %d - %d = %d" % (self.v1, self.v2, result))
        return result

    def setV1(self, v):
        if isinstance(v, int):
            self.v1 = v
    
    def getV1(self):
        return self.v1

    @classmethod
    def history(cls):
        for item in Cal._history:
            print(item)

class CalMultiply(Cal):
    def multiply(self):
        result = self.v1 * self.v2
        Cal._history.append("multiply : %d * %d = %d" % (self.v1, self.v2, result))
        return result

c1 = CalMultiply(10, 10)
print(c1.multiply())
c1.add()

Cal.history()

File: 인터넷 강의/python_ruby/Class_member/4.rb
class Cal
    @@_history = []

    def initialize(v1, v2)
        @v1 = v1
        @v2 = v2
    end
    #! @가 붙으면 인스턴스 변수이다.

    def add()
        result = @v1 + @v2
        @@_history.push("#{@v1}  + #{@v2} = #{result}")
        return result
    end

    def subtract()
        result = @v1 - @v2
        @@_history.push("#{@v1}  - #{@v2} = #{result}")
        return result
    end

    def setV1(v)
        if v.is_a?(Integer)
            @v1 = v
        end
    end

    def getV1()
        return @v1
    end

    def Cal.history()
        for item in @@_history
            p item
        end
    end

end

class CalMultiply < Cal
    def multiply()
        result = @v1 * @v2
        @@_history.push("#{@v1}  * #{@v2} = #{result}")
        return result
    end
end

c1 = CalMultiply.new(10,10)
p c1. add()
p c1. multiply
CalMultiply.history()